File: Patterns/Manuscript/manuscript/code/java/DesignPatterns/src/main/java/flyweight/Library.java
package flyweight;

import java.util.HashMap;
import java.util.Map;

class Library {
    private Map<String, Book> books = new HashMap<>();

    public Book getBook(String title, String author) {
        if (!books.containsKey(title)) {
            books.put(title, new ConcreteBook(title, author, books.size()+1));
        }
        return books.get(title);
    }

    public void borrowBook(String title, String author) {
        Book book = getBook(title, author);
        if (book.isAvailable()) {
            book.borrowBook();
        } else {
            // Simulate ordering a new copy
            Book newBook = new ConcreteBook(title, author, books.size()+1);
            books.put(title, newBook);
            System.out.println("New copy ordered for: " + title + " book id: " + (book.getId() +1) );

            newBook.borrowBook();
            books.put(title + " #" + (books.size() + 1), newBook); // Storing new copies distinctly
        }
    }

    public void returnBook(String title) {
        if (books.containsKey(title)) {
            books.get(title).returnBook();
        }
    }
}

File: code/csharp/DesignPatterns/flyweight/Library.cs
namespace Flyweight;
using System.Collections.Generic;

class Library
{
    private Dictionary<string, IBook> _books = new Dictionary<string, IBook>();

    public IBook GetBook(string title, string author)
    {
        if (!_books.ContainsKey(title))
        {
            _books[title] = new ConcreteBook(title, author, _books.Count + 1);
        }
        return _books[title];
    }

    public void BorrowBook(string title, string author)
    {
        IBook book = GetBook(title, author);
        if (book.IsAvailable())
        {
            book.BorrowBook();
        }
        else
        {
            IBook newBook = new ConcreteBook(title, author, _books.Count + 1);
            _books[title] = newBook;
            System.Console
            .WriteLine($"New copy ordered for: {title} book id: {book
                                                                .GetId() + 1}");
            newBook.BorrowBook();
        }
    }

    public void ReturnBook(string title)
    {
        if (_books.ContainsKey(title))
        {
            _books[title].ReturnBook();
        }
    }
}
